File: example/src/net/mm2d/testui/ActionNode.java
/*
 * Copyright(C) 2016 大前良介(OHMAE Ryosuke)
 */

package net.mm2d.testui;

import net.mm2d.upnp.Action;
import net.mm2d.upnp.Argument;
import net.mm2d.upnp.StateVariable;

import java.util.List;

/**
 * @author <a href="mailto:ryo@mm2d.net">大前良介(OHMAE Ryosuke)</a>
 */
public class ActionNode extends UpnpNode {
    public ActionNode(Action action) {
        super(action);
        final List<Argument> arguments = action.getArgumentList();
        for (final Argument argument : arguments) {
            add(new ArgumentNode(argument));
        }
    }

    @Override
    public String getDetailText() {
        final Action action = (Action) getUserObject();
        final StringBuilder sb = new StringBuilder();
        sb.append("Name: ");
        sb.append(action.getName());
        sb.append('\n');
        final List<Argument> args = action.getArgumentList();
        for (final Argument arg : args) {
            final StateVariable v = arg.getRelatedStateVariable();
            sb.append(arg.isInputDirection() ? "in:" : "out:");
            sb.append("(");
            sb.append(v.getDataType());
            sb.append(")");
            sb.append(arg.getName());
            sb.append('\n');
        }
        return sb.toString();
    }

    @Override
    public String toString() {
        final Action o = (Action) getUserObject();
        return o.getName();
    }
}

File: sample-cp/src/main/java/net/mm2d/upnp/sample/cp/ActionNode.kt
/*
 * Copyright (c) 2016 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp.sample.cp

import net.mm2d.upnp.cp.Action
import java.awt.Component
import javax.swing.JFrame
import javax.swing.JMenuItem
import javax.swing.JPopupMenu

/**
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class ActionNode(action: Action) : UpnpNode(action) {
    init {
        action.argumentList.forEach {
            add(ArgumentNode(it))
        }
    }

    override fun getUserObject(): Action {
        return super.getUserObject() as Action
    }

    override fun formatDescription(): String {
        return Formatter.format(getUserObject())
    }

    override fun toString(): String {
        return getUserObject().name
    }

    override fun showContextMenu(frame: JFrame, invoker: Component, x: Int, y: Int) {
        JPopupMenu().also {
            it.add(JMenuItem("Invoke Action").also { item ->
                item.addActionListener { ActionWindow(getUserObject()).show(frame.x + x, frame.y + y) }
            })
        }.show(invoker, x, y)
    }
}
